lib/hanami/providers/db.rb

```diff
@@ -9,23 +10,31 @@ module Providers
     class DB < Dry::System::Provider::Source
       extend Dry::Core::Cache
 
-      setting :database_url
+      include Dry::Configurable(config_class: Providers::DB::Config)
 
+      setting :database_url
       setting :adapter, default: :sql
+      setting :adapters, mutable: true, default: Adapters.new
+      setting :relations_path, default: "relations"
 
-      # TODO: Determine ideal default extensions
-      # TODO: Switch extensions based on configured adapter
-      setting :extensions, default: [:error_sql]
+      def initialize(...)
+        super(...)
 
-      setting :relations_path, default: "relations"
+        @configured_for_database = false
+      end
+
+      def finalize_config
```

Comment:
This is pulled out into a public method on the provider source so we can call it on the parent provider in `#apply_parent_config`. Doing it this way is important to avoid unnecessary connections to the database that would otherwise occurred if we called a straight-up `#prepare`.